sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/src/bot_description/bot_description/core.py

```diff
@@ -0,0 +1,166 @@
+from typing import List
+from collections import deque
+
+from promptulate.llms import ChatOpenAI, BaseLLM
+from promptulate.agents import BaseAgent
+from promptulate.output_formatter import OutputFormatter
+from promptulate.schema import MessageSet, SystemMessage, UserMessage, BaseMessage
+from promptulate.utils.color_print import print_text
+
+from bot_description.prompt import (
+    SYSTEM_PROMPT_TEMPLATE,
+    GENERATE_PLAN_SYSTEM_PROMPT,
+    ENVIRONMENT_SUMMARY_SYSTEM_PROMPT,
+    SECURITY_CHECK_SYSTEM_PROMPT,
+)
+from bot_description.schema import (
+    Task,
+    GeneratePlanResponse,
+    CommandResponse,
+    Operator,
+    Sensor,
+)
+
+
+class RobotController:
+    def __init__(self, operators: List[Operator]) -> None:
+        self.operators: List[Operator] = operators
+
+        if self.get_operator("stop") is None:
+            raise ValueError("Please provide stop operator")
+
+    def execute_action(self, action_name: str, *args, **kwargs) -> None:
+        """Execute action by action_name."""
+        operator = self.get_operator(action_name)
+        print_text(f"[command] run {action_name} {str(args)} {str(kwargs)}", "green")
+        operator.execute_action(*args, **kwargs)
+
+    def get_operator(self, action_name: str) -> Operator:
+        return next(
+            (operator for operator in self.operators if operator.name == action_name),
+            None,
+        )
+
+
+class RobotObserver:
+    def __init__(self, sensors: List[Sensor]) -> None:
+        self.sensors: List[Sensor] = sensors
+
+    def get_data(self, sensor_name: str) -> None:
+        sensor = self.get_sensor(sensor_name)
+        return sensor.get_data()
+
+    def get_sensor(self, sensor_name: str) -> Sensor:
+        return next(
+            (sensor for sensor in self.sensors if sensor.name == sensor_name),
+            None,
+        )
+
+    def show_all_data(self):
+        data = ""
```

Comment:
more Pythonic: `data = "".join([str(sensor.get_data()) for sensor in self.sensors])`